Mesh domain for MGXS

samuelshaner opened the pull request:
This PR adds a mesh domain option to the MGXS module. While the mesh domain option is currently limited to macroscopic cross sections due to the mesh not knowing which nuclides (or nuclide densities) lie within a given mesh cell, the utility of macroscopic cross section generation with a mesh domain is useful for kinetic calculations and likely for other uses, such as cross section generation for codes limited to cartesian meshes.

This PR builds on PR #680 and will greatly reduce in size once that one is merged.

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -7,13 +7,14 @@
 import openmc.checkvalue as cv
 from openmc.mgxs import EnergyGroups
-
+from openmc import Mesh
```
Any reason to import `Mesh` on its own rather than just use `openmc.Mesh` throughout?

samuelshaner replied:
Nope, and Mesh is never actually used like this. Must have needed it when somewhere in the development process. I'll take this out.

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -675,14 +684,20 @@ def load_from_statepoint(self, statepoint):
-        if self.domain_type != 'distribcell':
+        if self.domain_type == 'mesh':
+            filters = [self.domain_type]
+            xyz = map(lambda x: np.arange(1, x+1), self.domain.dimension)
```
PEP8 doesn't permit `lambda` expressions (though I personally like them!).

wbinventor replied:
Plus I think NumPy has some built-in routine to do this anyway.

samuelshaner replied:
I just took this line from `tallies.py`. I'll look into a solution that is PEP8 compliant and change it in both places.

wbinventor replied:
Oh ok. I'm personally fine with it, but perhaps @paulromano wants to take a hard line on lambdas.

paulromano replied:
The only thing PEP8 says is that you shouldn't assign a lambda. In this case, it's being used anonymously within a larger expression, so that would be fine per PEP8.

paulromano replied:
But yeah I'm pretty sure numpy has a better way to do this (combined with the next line)

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -1135,6 +1157,9 @@ def print_xs(self, subdomains='all', nuclides='all', xs_type='macro'):
         elif self.domain_type == 'distribcell':
             subdomains = np.arange(self.num_subdomains, dtype=np.int)
+        elif self.domain_type == 'mesh':
+            xyz = map(lambda x: np.arange(1, x+1), self.domain.dimension)
```
ditto lamda expressions - though i won't repeat this comment again.

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -4917,7 +4977,16 @@ def get_units(self, xs_type='macro'):
         """
 
-        cv.check_value('xs_type', xs_type, ['macro', 'micro'])
+        # Construct a collection of the subdomains to report
```
I'm a little confused - why is this block of code here?

wbinventor commented on tests/test_mgxs_library_hdf5/inputs_true.dat:
```diff
@@ -1 +1 @@
-e2cdca7ea5b3532050af5b12fac26d7ef212d2696bb1b73cdd00929b2243c40d100ad02438c7b090555b49815d0de6c48cf1b4ebf437a48bc80c2d2b4bad292e
\ No newline at end of file
+e2cdca7ea5b3532050af5b12fac26d7ef212d2696bb1b73cdd00929b2243c40d100ad02438c7b090555b49815d0de6c48cf1b4ebf437a48bc80c2d2b4bad292e
```
What changed in this test?

wbinventor commented:
This looks pretty good to me and should be a quick one to merge. Other than the lambda expressions, my only other question is whether we should include a test for MGXS on a mesh?

wbinventor commented:
Also, the Jupyter Notebooks should be updated to reflect the fact that mesh domains are now supported (right now they claim that materials, cells and universes are it). On a related note, have you confirmed that each of the Notebooks successfully runs with these code changes?

samuelshaner commented:
Thanks for the quick review @wbinventor. I addressed your code comments and replaced the lambda function with inline for loops which seems like a clean, quick modification.

At lease one test for MGXS is good idea - I'll develop one or two tests that I think would be useful.

The MGXS successfully run with the code changes, but in doing so I realized that I haven't added the mesh domain to the MGXS library so I'll have to go back and add that as well.

paulromano commented on openmc/mgxs/mgxs.py:
```diff
@@ -675,14 +683,20 @@ def load_from_statepoint(self, statepoint):
+        if self.domain_type == 'mesh':
+            filters = [self.domain_type]
+            xyz = [np.arange(1, x+1) for x in self.domain.dimension]
+            filter_bins = [tuple(itertools.product(*xyz))]
```
The numpy way to do this would be:
```Python
list(np.broadcast(*np.ogrid[1:x+1, 1:y+1, 1:z+1]))
```
Unfortunately you'd have to have separate cases for 2D and 3D, so in that sense I like yours better (I also think the meaning is clearer). However, I would suggest using just normal `range` instead of `np.arange` since at the end of the day you're getting a list of tuples.

paulromano commented on openmc/mgxs/mgxs.py:
```diff
@@ -756,12 +770,19 @@ def get_xs(self, groups='all', subdomains='all', nuclides='all',
+        # cells do not know the nuclide densities in each mesh cell.
+        if self.domain_type == 'mesh' and xs_type == 'micro':
+            msg = 'Unable to get micro xs for mesh domain since the mesh ' \
+                  'cells do not know the nuclide densities in each mesh cell.'
```
As it turns out, I'm actually working on a stochastic volume calculation capability right now, so we will be able to get around this soon.

paulromano replied:
And I should say -- the capability is both for getting the volume of an arbitrary region as well as average nuclide densities, the latter being what you need.

samuelshaner replied:
Cool, looking forward to see you how you do that.

wbinventor commented on docs/source/pythonapi/examples/mgxs-part-iv.ipynb:
```diff
@@ -519,9 +519,9 @@
    "metadata": {},
    "source": [
-    "Now we must specify the type of domain over which we would like the `Library` to compute multi-group cross sections. The domain type corresponds to the type of tally filter to be used in the tallies created to compute multi-group cross sections. At the present time, the `Library` supports \"material,\" \"cell,\" and \"universe\" domain types. In this simple example, we wish to compute multi- ...
+    "Now we must specify the type of domain over which we would like the `Library` to compute multi-group cross sections. The domain type corresponds to the type of tally filter to be used in the tallies created to compute multi-group cross sections. At the present time, the `Library` supports \"material\" \"cell\", \"universe\", and \"mesh\" domain types. In this simple example, we wish to compu ...
```
typo: "andtherefore" -> "and therefore

wbinventor commented on docs/source/pythonapi/examples/mgxs-part-iv.ipynb:
```diff
@@ -519,9 +519,9 @@
+    "Now we must specify the type of domain over which we would like the `Library` to compute multi-group cross sections. The domain type corresponds to the type of tally filter to be used in the tallies created to compute multi-group cross sections. At the present time, the `Library` supports \"material\" \"cell\", \"universe\", and \"mesh\" domain types. In this simple example, we wish to compu ...
     "\n",
-    "**Note:** By default, the `Library` class will instantiate `MGXS` objects for each and every domain (material, cell or universe) in the geometry of interest. However, one may specify a subset of these domains to the `Library.domains` property."
+    "**Note:** By default, the `Library` class will instantiate `MGXS` objects for each and every domain (material, cell, universe, or mesh cell) in the geometry of interest. However, one may specify a subset of these domains to the `Library.domains` property."
```
This is an OCD comment, so feel free to ignore it: based on my understanding of your implementation, "domain" refers to the "mesh" while "subdomain" refers to "mesh cell." Is this true? If so, perhaps consider revising "mesh cell" here to simply "mesh" unless there is something other reason you have for this wording choice (again, nbd, but these subtle word choices may lead to confusion if we aren't consistent).

samuelshaner replied:
That's correct. I'll clarify that here.

wbinventor commented on openmc/mgxs/library.py:
```diff
@@ -54,9 +54,10 @@ class Library(object):
+    domain_type : {'material', 'cell', 'distribcell', 'universe', 'mesh'}
         Domain type for spatial homogenization
-    domains : Iterable of openmc.Material, openmc.Cell or openmc.Universe
+    domains : Iterable of openmc.Material, openmc.Cell, openmc.Universe or
```
I don't think Sphinx will properly render this since it is split across two lines. Can you check that it shows up as expected?

samuelshaner replied:
Where would these lines show up in the sphinx documentation? When I build the docs with `make html` I don't see where this link would show up.

wbinventor replied:
Hmm I thought that the class attributes were picked up by Sphinx but it appears that is not rh case. So ignore my comment.

wbinventor commented on openmc/mgxs/library.py:
```diff
@@ -183,6 +184,8 @@ def domains(self):
             elif self.domain_type == 'universe':
                 return self.openmc_geometry.get_all_universes()
+            elif self.domain_type == 'mesh':
+                raise ValueError('Unable to get domains for Mesh domain type')
```
Should it not just return the mesh as the domain?

samuelshaner replied:
It could, but I don't see a way to get the mesh since the geometry doesn't have it. Do you think we should add the mesh as an attribute to the library?

wbinventor commented on openmc/mgxs/library.py:
```diff
@@ -273,6 +276,12 @@ def domains(self, domains):
             elif self.domain_type == 'universe':
                 cv.check_iterable_type('domain', domains, openmc.Universe)
                 all_domains = self.openmc_geometry.get_all_universes()
+            elif self.domain_type == 'mesh':
```
This is cool that you can compute MGXS on multiple meshes with the same library.

wbinventor commented on openmc/mgxs/mgxs.py:
```diff
@@ -1479,7 +1505,10 @@ def get_pandas_dataframe(self, groups='all', nuclides='all',
         # Remove the score column since it is homogeneous and redundant
-        df = df.drop('score', axis=1)
+        if self.domain_type == 'mesh':
+            df = df.drop('score', axis=1, level=0)
```
My understanding is that this line should work for other domain types as well.

samuelshaner replied:
If you specify a level, the axis must be a MultiIndex and the axis is only MultiIndex for a mesh domain.

wbinventor replied:
Oh I see. I wish Pandas had a slimmer API with a single family of index types, but oh well.

wbinventor commented:
I'm fine with this once my last few comments are addressed.

wbinventor commented:
Looks good!
